sotopia/experimental/agents/moderator.py

```diff
@@ -68,43 +68,143 @@ def __init__(
         self.current_agent_index: int = 0
         self.scenario: str = scenario
         self.agents: list[str] = list(agent_mapping.values())
+        self.agent_models: dict[str, str] = {}
+        self.agents_awake: dict[str, bool] = {name: False for name in self.agents}
+        self.all_agents_awake: asyncio.Event = asyncio.Event()
+        self.message_history: list[list[tuple[str, str, str]]] = [
+            [("Environment", "Environment", self.scenario)]
+        ]
 
     async def send(self, action: Observations) -> None:
         for output_channel, output_channel_type in self.output_channel_types.items():
-            await self.r.publish(
-                output_channel,
-                Message[output_channel_type](
-                    data=action.observations_map[output_channel]
-                ).model_dump_json(),  # type:ignore[valid-type]
-            )
+            if output_channel in action.observations_map:
+                await self.r.publish(
+                    output_channel,
+                    Message[output_channel_type](  # type:ignore[valid-type]
+                        data=action.observations_map[output_channel]
+                    ).model_dump_json(),
+                )
 
     async def __aenter__(self) -> Self:
         print(self.scenario)
-        await self.send(
-            Observations(
+        asyncio.create_task(self.booting())
+        self.task_scheduler = asyncio.create_task(self._task_scheduler())
+        return await super().__aenter__()
+
+    async def _task_scheduler(self) -> None:
+        await self.all_agents_awake.wait()
+        return await super()._task_scheduler()
+
+    async def booting(self) -> None:
+        """
+        1. send checking message to agents for every 5 seconds, until all agents are awake
+        - the agent should use this waking message to send their information to moderator for record
+        - if further information of the agents are needed, should be communicated through the booting process
+        2. after all agents are awake, send agent[0] a message to allow the agent to start speaking
+        """
+        while not self.all_agents_awake.is_set():
+            await self.send(
+                Observations(
+                    observations_map={
+                        output_channel: Observation(
+                            agent_name="moderator",
+                            last_turn=self.scenario,
```

Comment:
Please double check the sotopia implementation of the last turn msg here: https://github.com/sotopia-lab/sotopia/blob/61f190ee2c4b5acb94ef93c109225d6a474b4710/sotopia/envs/parallel.py#L300

The initial/boosting msg for every character should be different. This is a very important feature in Sotopia about info asymmetry

Reply:
I have made a temprary fix for this. I also notice that the current ScriptBackGround dataType only supports 2 agents in chat. I think I will create a new background dataType for moderator.

Reply:
thanks!